Implementation of SPLASH 1.0 in pyrealm (#69)
ImperialCollegeLondon/pyrealm: docs/source/users/splash.md

---
jupytext:
  formats: md:myst
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.13.8
kernelspec:
  display_name: pyrealm_python3
  language: python
  name: pyrealm_python3
---

# The `splash` submodule

The {mod}`~pyrealm.splash` module provides the SPLASH v1.0 model for estimating **soil

> moisture, actual evapotranspiration (AET) and soil water runoff** for a site

Reviewer: I think `surface runoff` instead of `soil water runoff` would be clearer here because I think SPLASH does not consider any below ground (=soil) water flows?

{cite:p}`davis:2017a`. The module in `pyrealm` completely reimplements the original
SPLASH code (see [the api notes](../api/splash_api.md) for details) but the outputs are
benchmarked against outputs from the original code in the `pyrealm` code testing.

The model takes an initial estimate of soil moisture and then uses time series of
precipitation, temperature and cloud cover to estimate how the daily water balance
changes with incoming precipitation , condensation and AET. The water balance
equation is:

> $W_{n[t]} = W_{n[t-1]} + P_{[t]} + C_{[t]} - \textrm{AET}_{[t]}$,

Reviewer: ```suggestion
$W_{n[t]} = W_{n[t-1]} + P_{[t]} + C_{[t]} - \textrm{AET}_{[t]} - RO$,
```
Did you leave out runoff on purpose here?

where the current soil moisture (mm, $W_{n[t]}$) is calculated from the previous day's
soil moisture (mm, $W_{n[t-1]}$),  given the expected AET (mm d-1,
$\textrm{AET}_{[t]}$), precipitation (mm d-1, $P_{[t]}$) and condensation (mm d-1,
$C_{[t]}$) for the current day, to calculate the current soil moisture (mm, $W_{n[t]}$).
Calculations of AET and condensation are affected by the soil moisture, temperature and
downwelling solar radiation at the site: this requires that the elevation and latitude
of the site are known.

The soil moisture at a site has a maximum value ($W_m$) that has a default value of
150mm. If the calculation above exceeds $W_m$, the excess water is allocated to runoff.

## Example data

The data below provides a 2 year daily time series of precipitation, temperature and

> solar fraction (1 - cloud cover) for 0.5° resolultion grid cells in a 10° by 10° block

Reviewer: ```suggestion
solar fraction (1 - cloud cover) for 0.5° resolution grid cells in a 10° by 10° block
```

::: {note}
The `SplashModel` code currently requires that the latitude (`lat`) and elevation
(`elv`) data have the same shape as the sunshine fraction (`sf`), temperature (`tc`) and
precipitation (`pn`). These values are obviously constant through time - and latitude
may well be constant across the longitude dimension for gridded data - but, at the
moment, you need to broadcast these variables to match.
:::

```{code-cell} python
splash = SplashModel(
    lat=np.broadcast_to(data.lat.data[None, :, None], data.sf.data.shape),
    elv=np.broadcast_to(data.elev.data[None, :, :], data.sf.data.shape),
    dates=Calendar(data.time.data),
    sf=data.sf.data,
    tc=data.tmp.data,
    pn=data.pre.data,
)
```

### Estimating initial soil moisture

In order to calculate water balances, you need initial values for the soil moisture.
This data is rarely available and so the
{meth}`~pyrealm.splash.splash.SplashModel.estimate_initial_soil_moisture` can be used
to estimate those values.

This method requires that the input data provides **at least one full year** of data. It
works by assuming that soil moisture is a **stationary** process on an annual time

> scale: the initial soil moisture should be the same as the soil moisture at the end of

Reviewer: I think `steady state` might be a better term here than `stationary process`. If you prefer to keep it I would change it to `soil moisture change is a stationary process`

moisture from the initial estimates (the 'previous day'). Note the saturated soil
moisture of 150mm near the coast and in the mountains.

```{code-cell} python
# Calculate the water balance equation for the first day from the initial soil
#  moisture estimates.
aet, wn, ro = splash.estimate_daily_water_balance(init_soil_moisture, day_idx=0)

# Plot the calculated soil moisture and change from previous values.
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(8, 4))
im_val = ax1.imshow(wn, extent=extent, origin="lower")
plt.colorbar(im_val, fraction=0.022, pad=0.03)
ax1.set_title("Soil moisture (mm)")

diff_val = ax2.imshow(init_soil_moisture - wn, extent=extent, origin="lower")
plt.colorbar(diff_val, fraction=0.022, pad=0.03)
ax2.set_title("Change in soil moisture (mm)")

plt.tight_layout()
```

The {meth}`~pyrealm.splash.splash.SplashModel.calculate_soil_moisture` method iterates
the daily estimation across all of the dates in the input data from initial soil
moisture estimates. It returns a set of time series of soil moisture, runoff and AET for
all sites.

```{code-cell} python
aet_out, wn_out, ro_out = splash.calculate_soil_moisture(init_soil_moisture)
```

> The plots below show the resulting soil moisture and Atime series for the three

Reviewer: ```suggestion
The plots below show the resulting soil moisture and a time series for the three
```
